File: Android/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/blockmonitor/core/MonitorCore.java
package com.didichuxing.doraemonkit.kit.blockmonitor.core;

import android.os.SystemClock;
import android.util.Printer;

import com.didichuxing.doraemonkit.kit.blockmonitor.bean.BlockInfo;

import java.util.ArrayList;

/**
 * @desc: 检测卡顿的日志类
 */
class MonitorCore implements Printer {
    private static final String TAG = "MonitorCore";
    /**
     * 卡顿阈值
     */
    private static final int BLOCK_THRESHOLD_MILLIS = 200;

    private long mStartTime = 0;
    private long mStartThreadTime = 0;
    private boolean mPrintingStarted = false;

    private StackSampler mStackSampler;


    public MonitorCore() {
        mStackSampler = new StackSampler();
        mStackSampler.init();
    }

    @Override
    public void println(String x) {
        if (!mPrintingStarted) {
            mStartTime = System.currentTimeMillis();
            mStartThreadTime = SystemClock.currentThreadTimeMillis();
            mPrintingStarted = true;
            mStackSampler.startDump();
        } else {
            final long endTime = System.currentTimeMillis();
            long endThreadTime = SystemClock.currentThreadTimeMillis();
            mPrintingStarted = false;
            if (isBlock(endTime)) {
                final ArrayList<String> entries = mStackSampler.getThreadStackEntries(mStartTime, endTime);
                if (entries.size() > 0) {
                    final BlockInfo blockInfo = BlockInfo.newInstance()
                            .setMainThreadTimeCost(mStartTime, endTime, mStartThreadTime, endThreadTime)
                            .setThreadStackEntries(entries)
                            .flushString();
                    BlockMonitorManager.getInstance().notifyBlockEvent(blockInfo);
                }
            }
            mStackSampler.stopDump();
        }
    }

    private boolean isBlock(long endTime) {
        return endTime - mStartTime > BLOCK_THRESHOLD_MILLIS;
    }


    public void shutDown() {
        mStackSampler.shutDown();
    }
}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/blockmonitor/core/MonitorCore.kt
package com.didichuxing.doraemonkit.kit.blockmonitor.core

import android.os.SystemClock
import android.util.Printer
import com.didichuxing.doraemonkit.kit.blockmonitor.bean.BlockInfo
import com.didichuxing.doraemonkit.kit.blockmonitor.core.BlockMonitorManager.Companion.instance

/**
 * @desc: 检测卡顿的日志类
 */
internal class MonitorCore : Printer {
    private var mStartTime: Long = 0
    private var mStartThreadTime: Long = 0
    private var mPrintingStarted = false
    private val mStackSampler: StackSampler = StackSampler()

    override fun println(x: String) {
        if (!mPrintingStarted) {
            mStartTime = System.currentTimeMillis()
            mStartThreadTime = SystemClock.currentThreadTimeMillis()
            mPrintingStarted = true
            mStackSampler.startDump()
        } else {
            val endTime = System.currentTimeMillis()
            val endThreadTime = SystemClock.currentThreadTimeMillis()
            mPrintingStarted = false
            if (isBlock(endTime)) {
                val entries = mStackSampler.getThreadStackEntries(mStartTime, endTime)
                if (entries.size > 0) {
                    val blockInfo = BlockInfo.newInstance()
                            .setMainThreadTimeCost(mStartTime, endTime, mStartThreadTime, endThreadTime)
                            .setThreadStackEntries(entries)
                            .flushString()
                    instance.notifyBlockEvent(blockInfo)
                }
            }
            mStackSampler.stopDump()
        }
    }

    private fun isBlock(endTime: Long): Boolean {
        return endTime - mStartTime > BLOCK_THRESHOLD_MILLIS
    }

    fun shutDown() {
        mStackSampler.shutDown()
    }

    companion object {
        private const val TAG = "MonitorCore"

        /**
         * 卡顿阈值
         */
        private const val BLOCK_THRESHOLD_MILLIS = 200
    }

    init {
        mStackSampler.init()
    }
}